Fix association on junction table between the same model.

KeitaMoromizato opened the pull request:
This pull request fix association on junction table between the same model.

example:

``` node
// user.js
module.exports = {
  tableName: 'users',
  attributes: {
    name: 'string',
    follows: {
      collection: 'users',
      through: 'follow',
      via: 'from'
    },
    followers: {
      collection: 'users',
      through: 'follow',
      via: 'to'
    }
}

// follow.js
module.exports = {

  tableName: 'follows',
  tables: ['users', 'users'],
  junctionTable: true,

  attributes: {
    to: {
      columnName: 'to',
      foreignKey: true,
      references: 'user',
      on: 'id',
      onKey: 'id',
      via: 'from'
    },
    from: {
      columnName: 'from',
      foreignKey: true,
      references: 'user',
      on: 'id',
      onKey: 'id',
      via: 'to'
    }
  }
};
```

The current version always use `via to` association.

devinivy commented:
Mind writing a failing test for this?

KeitaMoromizato commented:
added test code.

KeitaMoromizato commented:
@devinivy 
Sorry, description and test code was not enough. I wrote test code and Details of this Pull Request below.


I want to use junction table like follow/followers. so I tried following code. 

``` node
// user.js
module.exports = {
  attributes: {
    name:  'string',
    follows: {
      collection: 'follow',
      references: 'follow',
      on: 'to',
      via: 'from'
    },
    followers: {
      collection: 'follow',
      references: 'follow',
      on: 'from',
      via: 'to'
    }
}

// follow.js
module.exports = {

  tableName: 'follows',
  tables: ['users', 'users'],
  junctionTable: true,

  attributes: {
    to: {
      columnName: 'to',
      type: 'integer',
      foreignKey: true,
      references: 'user',
      on: 'id',
      groupKey: 'user'
    },
    from: {
      columnName: 'from',
      type: 'integer',
      foreignKey: true,
      references: 'user',
      on: 'id',
      groupKey: 'user'
    }
  }
};
```

But `User.find().populate('follows')` return following. I expect  User Collection in User.follows ( like `follows[ { id: 1, name: 'huga'}]`).

```
{
  id: 1,
  name: 'hoge',
  follows: [
    {
      id: 1,
      to: 1,
      from: 2
    }
  ]
}
```

so I created this Pull Request for solve this problem. If this method and Pull Request is incorrect,  Would you mind teaching to me how to solve this problem ?

devinivy commented:
I'll take a peek at this soon!  It looks like you're trying to use through associations, which are currently a little finicky.  Through associations are working for me, and same-table many-to-many relationships are working for me, but I haven't used both together.  See discussion/example of through-associations here: https://github.com/balderdashy/waterline/issues/705#issuecomment-60945411

dmarcelino commented:
I like this, this actually fixes a couple of many-to-many through self-referencing tests in `sails-orientdb`: https://github.com/appscot/sails-orientdb/blob/master/test/integration-orientdb/tests/associations/manyThroughSelf.add.js#L78

I don't see any obvious issue with the code change (but I'm no waterline-schema expert) so if @devinivy is happy with them I say lets merge this. :white_check_mark:

devinivy commented:
I need to figure out what this does, and what's failing.  I don't quite understand what this fixes yet.  That said, I've not tested using both through-associations and self-referencing many-to-many relationships together.  They both work separately.

dmarcelino commented:
@devinivy, I was on vacation and only seen this now. Those failing tests are not in waterline-adapter-tests because currently many-to-many through is not supported. In my tests' fixtures I'm using your hack/workaround from https://github.com/balderdashy/waterline/issues/705#issuecomment-60945411 because we *kind of* promote it in sails-orientdb.

devinivy commented on test/joinTables.js:
```diff
@@ -373,4 +373,71 @@ describe('JoinTables', function() {
+        follows: {
+          collection: 'foo',
+          through: 'bar',
+          via: 'from'
```
Is this line and line 402 supposed to be the same?  Shouldn't one of the `via` values be `'to'`?

devinivy commented on test/joinTables.js:
```diff
@@ -373,4 +373,71 @@ describe('JoinTables', function() {
+    it('should join to collect attributes', function() {
+      var obj = new JoinTables(collections);
+
+      assert(obj.foo.follows.onKey !== 'from');
```
Could you make these assertions using `assert.equal`, and assert the actual value rather than the non-value?

devinivy commented:
Sorry this has taken so long!  I'd love to see this go through.  Made a couple line notes!

KeitaMoromizato commented:
Thanks!

> Is this line and line 402 supposed to be the same? Shouldn't one of the via values– most likely this one– be 'to'?

Sorry. It's typo. I fixed it!

> Could you make these assertions using assert.equal, and assert the actual value rather than the non-value?

I modified to use `assert.equal()`, and fixed bug. I had mistake on definition of attributes...

devinivy commented:
Great!  I'm okay with this change.  The new couple lines of code ensure that, when looking for a reference, it's checked against both `collection` and `via` rather than just `collection`.  This basically avoids confusion when a junction table has multiple attributes referencing the same collection, as in the case of a simple through table in which a collection references itself.

dmarcelino commented:
Thanks @KeitaMoromizato!
